test/common/network/apple_dns_impl_test.cc

```diff
@@ -194,6 +179,499 @@ TEST_F(AppleDnsImplTest, Timeout) {
+  addr4_2.sin_port = htons(6502);
+  Network::Address::Ipv4Instance address2(&addr4);
+
+  DNSServiceGetAddrInfoReply reply_callback;
+  absl::Notification dns_callback_executed;
+
+  EXPECT_CALL(dns_service_,
+              dnsServiceGetAddrInfo(_, kDNSServiceFlagsShareConnection | kDNSServiceFlagsTimeout, 0,
+                                    kDNSServiceProtocol_IPv4 | kDNSServiceProtocol_IPv6,
+                                    StrEq(hostname.c_str()), _, _))
+      .WillOnce(DoAll(SaveArg<5>(&reply_callback), Return(kDNSServiceErr_NoError)));
+
+  auto query =
+      resolver_->resolve(hostname, Network::DnsLookupFamily::Auto,
+                         [&dns_callback_executed](DnsResolver::ResolutionStatus status,
+                                                  std::list<DnsResponse>&& response) -> void {
+                           EXPECT_EQ(DnsResolver::ResolutionStatus::Success, status);
+                           EXPECT_EQ(2, response.size());
+                           dns_callback_executed.Notify();
+                         });
+  ASSERT_NE(nullptr, query);
+
+  // Fill the query with one address, and promise more addresses are coming. Meaning the query will
+  // be pending.
+  reply_callback(nullptr, kDNSServiceFlagsAdd | kDNSServiceFlagsMoreComing, 0,
+                 kDNSServiceErr_NoError, hostname.c_str(), address.sockAddr(), 30, query);
+
+  // The query's ref is going to be deallocated when the query is destroyed.
+  EXPECT_CALL(dns_service_, dnsServiceRefDeallocate(_));
+  reply_callback(nullptr, kDNSServiceFlagsAdd, 0, kDNSServiceErr_NoError, hostname.c_str(),
+                 address2.sockAddr(), 30, query);
+
+  dns_callback_executed.WaitForNotification();
+}
+
+TEST_F(AppleDnsImplFakeApiTest, MultipleAddressesSecondOneFails) {
+  createResolver();
+
+  const std::string hostname = "foo.com";
+  sockaddr_in addr4;
+  addr4.sin_family = AF_INET;
+  EXPECT_EQ(1, inet_pton(AF_INET, "1.2.3.4", &addr4.sin_addr));
+  addr4.sin_port = htons(6502);
+  Network::Address::Ipv4Instance address(&addr4);
+
+  DNSServiceGetAddrInfoReply reply_callback;
+  absl::Notification dns_callback_executed;
+
+  EXPECT_CALL(dns_service_,
+              dnsServiceGetAddrInfo(_, kDNSServiceFlagsShareConnection | kDNSServiceFlagsTimeout, 0,
+                                    kDNSServiceProtocol_IPv4 | kDNSServiceProtocol_IPv6,
+                                    StrEq(hostname.c_str()), _, _))
+      .WillOnce(DoAll(SaveArg<5>(&reply_callback), Return(kDNSServiceErr_NoError)));
+
+  auto query =
+      resolver_->resolve(hostname, Network::DnsLookupFamily::Auto,
+                         [&dns_callback_executed](DnsResolver::ResolutionStatus status,
+                                                  std::list<DnsResponse>&& response) -> void {
+                           EXPECT_EQ(DnsResolver::ResolutionStatus::Failure, status);
+                           // DISCUSSION: should the responses be cleared because of failure?
```

Comment:
historically we have let the callback target decide what to do with a Failure status.

Reply:
Can you expand?

Reply:
Currently, the query might have accumulated addresses in previous replies from the API before the error occurs. Therefore, when the query is flushed and the ResolveCb is called the resolver would pass a Failure status, but with non-empty response list. 

Currently all ResolveCb targets code defensively by ignoring the response list if the status is a failure. But I was wondering if we should be extra-defensive here and clear the response list when failure is detected, so that a ResolveCb target might not be "tempted" to use potential addresses received when the resolution status is a failure.

Reply:
OK I see. Yeah, clearing can't hurt.